### ISSUE TITLE
Concurrency Issue While running MOEA Algorithms

### ISSUE THREAD

**Issue by mboussaa** [MOEAFramework/MOEAFramework#35]
Hi,

I am running a sample NSGAII algorithm. 
The "evaluate" function responsable for evaluating each new generated individual in the population open a remote connection to the cloud (docker client to the cloud) to run my staff and get back the output value (fitness value).

Since that, all things are going well. However, while the algorithm is evaluating individuals, sometimes it get stuck in some evaluations by triggering an exception 
```
com.spotify.docker.client.DockerRequestException: Request error: POST http://10.0.0.22:2375/v1.12/containers/create?name=execution_container_consumer_tiffmedian_: 409
com.spotify.docker.client.DockerRequestException: Request error: POST http://10.0.0.22:2375/v1.12/containers/create?name=execution_container_consumer_tiffmedian_: 409
	at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:1061)
	at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:1010)
	at com.spotify.docker.client.DefaultDockerClient.createContainer(DefaultDockerClient.java:370)
	at fr.inria.diverse.geneticAlgorithms.multi.Processing.executeContainers(Processing.java:110)
	at fr.inria.diverse.geneticAlgorithms.multi.TimeMemoryProblem.evaluate(TimeMemoryProblem.java:28)
	at org.moeaframework.algorithm.AbstractAlgorithm.evaluate(AbstractAlgorithm.java:98)
	at org.moeaframework.algorithm.AbstractAlgorithm.evaluateAll(AbstractAlgorithm.java:82)
	at org.moeaframework.algorithm.NSGAII.iterate(NSGAII.java:90)
	at org.moeaframework.algorithm.AbstractAlgorithm.step(AbstractAlgorithm.java:161)
	at org.moeaframework.Executor.runSingleSeed(Executor.java:700)
	at org.moeaframework.Executor.run(Executor.java:629)
	at fr.inria.diverse.geneticAlgorithms.multi.NSGAII.NSGAII.main(NSGAII.java:55)
Caused by: javax.ws.rs.ClientErrorException: HTTP 409 Conflict
	at org.glassfish.jersey.client.JerseyInvocation.createExceptionForFamily(JerseyInvocation.java:991)
	at org.glassfish.jersey.client.JerseyInvocation.convertToException(JerseyInvocation.java:975)
	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:795)
	at org.glassfish.jersey.client.JerseyInvocation.access$500(JerseyInvocation.java:91)
	at org.glassfish.jersey.client.JerseyInvocation$5.completed(JerseyInvocation.java:756)
	at org.glassfish.jersey.client.ClientRuntime.processResponse(ClientRuntime.java:189)
	at org.glassfish.jersey.client.ClientRuntime.access$300(ClientRuntime.java:74)
	at org.glassfish.jersey.client.ClientRuntime$1.run(ClientRuntime.java:171)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:320)
	at org.glassfish.jersey.client.ClientRuntime$2.run(ClientRuntime.java:201)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

If I run my algorithm for 100 evaluations that happens for at least 5 or 6 times.

It seems like MOEA is trying to run parallel evaluations. I am not sure.

```
NondominatedPopulation result = new Executor()
				.withProblemClass(c)
				.withAlgorithm("NSGAII")
				.withMaxEvaluations(50)
				.withProperty("populationSize", 20)
				.withProperty("1x.rate", 0.5)
				.withProperty("bf.rate", 0.1)
				.run();
```
As well, I didn't ask NSGA II to run on multiple threads or distribute workload on all cores

**Comment by dhadka** [comment]
The MOEA Framework shouldn't run in parallel unless you specifically enable it.  It could have something to do with the REST library (Jersey) running asynchronous queries.  You could try adding the `synchronized ` keyword to the evaluate method to guarantee it does not run in parallel:

```java

public synchronized void evaluate(Solution solution) {
      ....
}
```

Without knowing what's causing the 409 Conflict, it's hard to tell what's going on.  While it doesn't solve the specific issue, you could catch the exception and retry the POST.

**Comment by dhadka** [comment]
Closing issue due to inactivity.  Please reopen or create a new issue if you have additional questions.
